starlette/_utils.py

```diff
@@ -0,0 +1,12 @@
+import asyncio
+import functools
+import typing
+
+
+def iscoroutinefunction(obj: typing.Any) -> bool:
+    while isinstance(obj, functools.partial):
```

Comment:
You might want to handle functools.wraps as well (you can check obj.wrapped)

Reply:
What case does it cover?

Reply:
Something like:

```python
import functools

db = object()

def inject_db(func):
    @functoolswraps(func)
    def wrapper(req):
        return func(req, db)
    return wrapper

@inject_db
def endpoint(request, db):
    ...
```

Reply:
I'll not be doing this on this PR, as there's quite a big discussion on #734, and it didn't attract much attention on those years... But doing that check does solve #734.

Reply:
No problem, we can easily do it in a followup